integration_tests/tests/mod.rs

```diff
@@ -0,0 +1,408 @@
+            // Expiry is either:
+            // - the same as customer close, but preceded by the customer noticing expiry
+            // and posting the corrected balances (+70)
+            // - the merchant waits for self-delay to elapse, then claims funds (130)
+            Self::MerchantExpiry => 200,
+        };
+        Duration::from_secs(seconds)
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct Outcome {
+    /// Which process, if any, had an error? Assumes that exactly one party will error.
+    error: Option<Party>,
+    /// Outcome of channel; left an Option in case we do not expect a channel from a test
+    /// (e.g. interacting with a non-existent channel).
+    channel_outcome: Option<ChannelOutcome>,
+}
+
+#[derive(Debug, Clone)]
+struct ChannelOutcome {
+    customer_status: CustomerStatus,
+    merchant_status: MerchantStatus,
+    customer_balance: CustomerBalance,
+    merchant_balance: MerchantBalance,
+    /// Closing balances of channels, for use in close-related tests
+    closing_balances: Option<ClosingBalances>,
+}
+
+#[derive(Debug, Error)]
+enum TestError {
+    #[error("Operation {0:?} not yet implemented")]
+    NotImplemented(Operation),
+
+    #[error(
+        "The error behavior did not satisfy expected behavior {op:?}. Got
+    CUSTOMER WATCHER OUTPUT:
+    {customer_errors}
+    MERCHANT SERVER OUTPUT:
+    {merchant_errors}
+    OPERATION OUTPUT:
+    {op_error:?}"
+    )]
+    InvalidErrorBehavior {
+        op: Operation,
+        customer_errors: String,
+        merchant_errors: String,
+        op_error: Result<(), anyhow::Error>,
+    },
+
+    #[error("After {op:?}, expected customer status {expected:?}, got {actual:?}")]
+    InvalidCustomerStatus {
+        op: Operation,
+        expected: CustomerStatus,
+        actual: CustomerStatus,
+    },
+
+    #[error("After {op:?}, expected merchant status {expected:?}, got {actual:?}")]
+    InvalidMerchantStatus {
+        op: Operation,
+        expected: MerchantStatus,
+        actual: MerchantStatus,
+    },
+
+    #[error(
+        "After {op:?}, expected customer, merchant balances
+    ({expected_customer:?}, {expected_merchant:?}), got
+    ({actual_customer:?}, {actual_merchant:?})"
+    )]
+    InvalidChannelBalances {
+        op: Operation,
+        expected_customer: CustomerBalance,
+        expected_merchant: MerchantBalance,
+        actual_customer: CustomerBalance,
+        actual_merchant: MerchantBalance,
+    },
+
+    #[error(
+        "After {op:?}, party: {party:?} expected customer, merchant closing balances
+    ({expected_customer:?}, {expected_merchant:?}), got
+    ({actual_customer:?}, {actual_merchant:?})"
+    )]
+    InvalidClosingBalances {
+        op: Operation,
+        party: String,
+        expected_customer: Option<CustomerBalance>,
+        expected_merchant: Option<MerchantBalance>,
+        actual_customer: Option<CustomerBalance>,
+        actual_merchant: Option<MerchantBalance>,
```

Comment:
suggestion: You can use the `ClosingBalances` struct in the error directly.

Reply:
I changed the test to do this, but it required me to clone the values to put into the error. Can revert that if that's not something we want.

Reply:
The clone is fine, but I also think it's okay to derive `Copy` for `ClosingBalances`. The struct itself is just two `u64`s with a bunch of layers of type wrappers around it. `Copy` is for types that are suitable to call `memcpy` on, and I think it's fine here.

Reply:
Went with deriving `Copy`, it looked clearer